### PR TITLE
Protocol compat

### PR DESCRIPTION
**Pull request by itsjamie**
Added on to the compatibility checks based on #EXT-X-VERSION.

In addition, I added the functionality that the `{Key|Map}` struct is copied and appended to every segment that it applies to. Therefore, I added the detection on encoding to only print when the previous segment was different.

### CODE REVIEW

**Comment by bradleyfalzon** [comment]
Note, internally we found a device (I know, we can't hack around shitty devices and I think it was an obscure STB), that couldn't handle reusing keys between segments without being explicitly told to.

We ended up dealing with it by creating a concept for output filters that we could chain on demand.

I'm not sure how this manifest package handles things right now, but introducing this concept if it's not already there may help separating RFC compliant behaviour from other formatting possibilities. 

For the record, we also had filters to add http headers, limiting live stream segments (forcing clients to start a live stream at an exact position as well as forcing delays if needed), standardising KeyFormatVersions, stripping SCTE tags, ConsistentKeys (what I just mentioned) and forcing NONE encryption (when switching between playlists with and without encryption).

As an example, we had a filters.go file which defined all the possible filters (along with filter_test.go to test them). Our implementation was tied to the HTTP handlers, so we could add headers, but each filter was created like so:

```go
  9 // MediaFilter is ran before a playlist and http response is sent to client
 10 type MediaFilter interface {
 11     Filter(*m3u8.MediaPlaylist, http.ResponseWriter, *http.Request) http.ResponseWriter
 12 }
 13
 14 // AddMediaFilter adds a filter to be ran at a later time
 15 func (s *Stream) AddMediaFilter(f MediaFilter) {
 16     s.mediaFilters = append(s.mediaFilters, f)
 17 }

111 // ConsistentKeys adds aes keys to each segment, some devices (safari/quicktime)
112 // were having issues when a series of segments that used the same key/iv but
113 // when it was only defined once, the client would crash. This sets those same
114 // keys each time.
115 type ConsistentKeys struct{}
116
117 // Filter implements MediaFilter interface
118 func (f ConsistentKeys) Filter(pls *m3u8.MediaPlaylist, w http.ResponseWriter, r *http.Request) http.ResponseWriter {
119     var lastKey *m3u8.Key
120     for i := uint(0); i < pls.Count(); i++ {
121         if pls.Segments[i] == nil {
122             break
123         }
124         if pls.Segments[i].Key == nil {
125             pls.Segments[i].Key = lastKey
126         }
127         lastKey = pls.Segments[i].Key
128     }
129     return w
130 }
```

This  makes it easy to write comprehensive tests for complicated cases. And we'd just append these to a `[]Filter`

And played like so:

```go
319     stream.AddMediaFilter(ConsistentKeys{})
320     stream.AddMediaFilter(LimitLiveFirstPlayChunks{Limit: 3, PlayStartMSN: stream.PlayStartMSN})
321     w = stream.RunMediaFilters(content, w, r)
322
323     fmt.Fprint(w, content.String())
```

`manifest` package probably doesn't care about http reader/writer component, but the rewriting of the manifest is what I was trying to illustrate.

File: hls/decode-util_test.go [hunk 1]
```diff
@@ -2,37 +2,101 @@ package hls
...
+			expectErr: false,
+			mediaType: "master",
+			file:      "fixture-v7.m3u8",
+		},
 	}
 
 	for _, tt := range tests {
 		t.Run(tt.file, func(t *testing.T) {
-			f, err := os.Open("./testdata/" + tt.file)
+			f, err := os.Open(fmt.Sprintf("./testdata/%s/%s", tt.mediaType, tt.file))
```
Comment (bradleyfalzon) [inline hls/decode-util_test.go]: Could probably use https://golang.org/pkg/path/filepath/#Join here if you liked.

Comment (itsjamie) [reply hls/decode-util_test.go]: Thanks, made that change for any `\` souls out there 👍

**Comment by itsjamie** [comment]
@bradleyfalzon the filter concept seems useful and easy to compose manipulation. I'll add an issue to design the interface for this package.

I'm concerned but not surprised that you found a device that required the key be provided for every segment.

**Review by tainacleal — DISMISSED** [review]

File: hls/encode-util.go [hunk 2]
```diff
@@ -205,18 +205,36 @@ func (p *MediaPlaylist) writeIFramesOnly(buf *manifest.BufWrapper) {
...
+			if previousSegment != nil {
+				for _, oldKey := range previousSegment.Keys {
+					if key == oldKey {
+						found = true
+						break
+					}
 				}
+			} else {
+				found = true
+			}
```
Comment (tainacleal) [inline hls/encode-util.go]: Got a quick question about this else block, why is "found" being set to true if there isn't a previous segment?

Comment (itsjamie) [reply hls/encode-util.go]: That would be... uh.. this really complicated thing.. that uh..

💯

**Review by tainacleal — APPROVED** [review]

**Pull request merged**